ksql-rest-app/src/test/resources/ksql-plan-schema/schema-final.json

```diff
@@ -0,0 +1,885 @@
+{
+  "$schema" : "http://json-schema.org/draft-04/schema#",
+  "title" : "Ksql Plan",
+  "oneOf" : [ {
+    "$ref" : "#/definitions/KsqlPlanV1"
+  } ],
+  "definitions" : {
+    "KsqlPlanV1" : {
+      "type" : "object",
+      "additionalProperties" : false,
```

Comment:
Continuing a discussion from #3520 :

> As above, no allowing additional properties may well cause us pain later.

> This is what we want. If there's an unrecognized field, that means that some newer version wrote that field and will probably use it when building the final query plan. We therefore don't want the current version to build the plan, because it likely doesn't have enough information to build it correctly. The goal here is to be backwards-compatible. New versions should be able to read and execute old plans, which this permits as long as we only add properties to the schema (we should extend our schema generation/checking to validate this - I'm looking into options here).

Reply:
First, a small nit: we should name this `ignoreUnknown` to be consistent with jackson annotations.

> The goal here is to be backwards-compatible. New versions should be able to read and execute old plans, which this permits as long as we only add properties to the schema (we should extend our schema generation/checking to validate this - I'm looking into options here).

I suggested using AVRO as the serialization format in an original comment on the KLIP, but the more I think about it the more value I think it has. AVRO has a very clear schema compatibility and validation mechanism that we can leverage - and the concept of reader/writer schemas allows schema evolution in a really clean way. It also is much easier to implement what @big-andy-coates suggested to split up the schema into smaller files. Not to mention efficiency...

> I'm not sure of the right answer. However, my first reaction would be that in general each version of an object's schema should be immutable. However, maybe this is overly restrictive and will lead to a blooming of unnecessary versions.

I agree that this is too restrictive. For the most part, I should be able to add new fields as long as they have defaults and I don't change the meaning of old ones (Avro can also test this for you).

Reply:
> First, a small nit: we should name this ignoreUnknown

`additionalProperties` is a part of the standard JSON schema...

Reply:
The avro idea is interesting, though I think I hear @rodesai groaning.  This kind of decision should probably have been made earlier in the process, e.g. as part of the KLIP review! 

Though we are were we are and it'll be a hell of a lot less work to make the change to Avro now than in 6 months once the JSON stuff has been released.

If we did go down the Avro route then I wouldn't want a hard dependency on SR. So we'd either need to publish the schema as part of each message, or have each KSQL release include all previous schema versions. (Remember you _must_ have the exact write schema when deserializing Avro).

@rodesai - what's your thoughts?

Reply:
> How does this play with adding new versions of an object? i.e. when do you see us adding a new field to an existing object version vs creating a new version of an object? This is currently unclear to me, (and maybe should be documented somewhere).

The intent here is what you guys have described above:

Developers should be able to change step schemas in a (transitively) backwards-compatible way by adding fields or making existing fields optional, and ensuring that the engine executes previous generations of a given's step the exact same way. We will verify this by:
    - verifying schema changes are backwards compatible.
    - extending the current query validation tests to verify that KSQL executes older plans the same way. We can do this by running query validation on all generations of the plan for a given query, and failing the test every time KSQL detects the plan for a given query has changed. When this happens, the developer has to write out a new plan (alongside the old plans) for the query.

Steps get version bumps if we want to completely change how a step is executed.

Reply:
Filed https://github.com/confluentinc/ksql/issues/4221 to cover adding documentation about evolving the schema